src/core/metrics.cc

```diff
@@ -172,20 +172,42 @@ Metrics::InitializeNvmlMetrics()
     return false;
   }
 
-  unsigned int dcnt;
-  nvmlerr = nvmlDeviceGetCount(&dcnt);
-  if (nvmlerr != NVML_SUCCESS) {
-    LOG_ERROR << "failed to get device count for nvml metrics: NVML_ERROR "
-              << nvmlerr;
+  int dcnt;
+  cudaError_t cudaerr = cudaGetDeviceCount(&dcnt);
+  if (cudaerr != cudaSuccess) {
+    LOG_ERROR << "failed to get device count for cuda devices: CUDA_ERROR "
+              << cudaerr;
     return false;
   }
 
   // Create NVML metrics for each GPU
-  LOG_INFO << "found " << dcnt << " GPUs supporting NVML metrics";
-  for (unsigned int didx = 0; didx < dcnt; ++didx) {
+  LOG_INFO << "found " << dcnt << " CUDA visible devices.";
+  for (int didx = 0; didx < dcnt; ++didx) {
     // Get handle for the GPU
+    cudaDeviceProp gpu_properties;
+    cudaError_t cudaerr = cudaGetDeviceProperties(&gpu_properties, didx);
+    if (cudaerr != cudaSuccess) {
+      LOG_ERROR << "failed to get device properties for cuda device " << didx << "CUDA_ERROR "
+                << cudaerr;
+      return false;
+    }
+
+    char pcibusid_str[64];
+    cudaError_t cuerr = cudaDeviceGetPCIBusId(
+        pcibusid_str, sizeof(pcibusid_str) - 1, didx);
+    if (cuerr != cudaSuccess) {
+      LOG_ERROR << "failed to get PCI Bus ID for CUDA device " << didx
+                << ": " << cudaGetErrorString(cuerr);
+      return false;
+    }
+
     nvmlDevice_t gpu;
-    nvmlReturn_t nvmlerr = nvmlDeviceGetHandleByIndex(didx, &gpu);
+    nvmlReturn_t nvmlerr = nvmlDeviceGetHandleByPciBusId_v2(pcibusid_str, &gpu);
+    if (nvmlerr != NVML_SUCCESS) {
+      LOG_ERROR << "failed to get device from PCI Bus ID: NVML_ERROR " << nvmlerr;
```

Comment:
You can remove the NVML_ERROR and use `nvmlErrorString(nvmlerr)` instead. Look [here](https://docs.nvidia.com/deploy/nvml-api/group__nvmlErrorReporting.html#group__nvmlErrorReporting_1g93eca32cce1cbd5961fbecce10d41750).